lightning-invoice/src/payment.rs

```diff
@@ -399,20 +465,29 @@ where
 	fn retry_payment(
 		&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters
 	) -> Result<(), ()> {
-		let max_payment_attempts = self.retry_attempts.0 + 1;
 		let attempts = *self.payment_cache.lock().unwrap()
 			.entry(payment_hash)
-			.and_modify(|attempts| *attempts += 1)
-			.or_insert(1);
+			.and_modify(|attempts| {
+				attempts.retry_count += 1;
+			})
+			.or_insert(PaymentAttempts::new());
+
+		let prevoius_retry_count = match attempts.retry_count {
```

Comment:
Why not just do the `retry_count` increment after checking `is_retryable_now`?

Reply:
I will give it another try this weekend(Due to my poor knowledge in Rust hashmap I was not able to write sth not too clumsy)

Reply:
https://github.com/lightningdevkit/rust-lightning/pull/1418/commits/ca5de8759f213b8ae720faed39c6731fca878f9b

I have separated the change to a separate PR for easier review and I will squash them if it looks good to you. I was trying to only calling `self.payment_cache.lock().unwrap()` and assigned it to a mut var once but this approach end up in dead lock in test so I just keep it simpler like the existing approach.

Reply:
Could you have another look to see if the current approach is reasonable? Sorry that this simple PR result in a lot of back-and-forth review @jkczyz @TheBlueMatt